sbi/src/pmu.rs

```diff
@@ -0,0 +1,823 @@
+        PmuCounterConfigFlags(self.0 | (1 << 6))
+    }
+
+    /// Returns if the sinh bit-flag is set.
+    pub fn get_sinh(&self) -> bool {
+        self.0 & (1 << 6) != 0
+    }
+
+    /// Sets the minh bit-flag (inhibit counter in M-mode).
+    pub fn set_minh(self) -> Self {
+        PmuCounterConfigFlags(self.0 | (1 << 7))
+    }
+
+    /// Returns if the minh bit-flag is set.
+    pub fn get_minh(&self) -> bool {
+        self.0 & (1 << 7) != 0
+    }
+}
+
+/// This encapsulates the bit-fields for PMU start_flags parameter as described in the SBI documentation
+/// for sbi_pmu_counter_start
+#[derive(Copy, Clone, Debug, Default)]
+pub struct PmuCounterStartFlags(u64);
+
+impl PmuCounterStartFlags {
+    /// Constructs a new PmuCounterStartFlags from a valid passed-in value.
+    pub fn from_raw_value(value: u64) -> Result<Self> {
+        match value {
+            0 | 1 => Ok(PmuCounterStartFlags(value)),
+            _ => Err(Error::InvalidParam),
+        }
+    }
+
+    /// Returns the raw value of the inner bit-flag field.
+    pub fn raw(&self) -> u64 {
+        self.0
+    }
+
+    /// Sets the set_init_value bit-flag (set initial counter value).
+    pub fn set_init_value(self) -> Self {
+        PmuCounterStartFlags(self.0 | 1)
+    }
+
+    /// Returns if set_init_value bit-flag is set.
+    pub fn get_init_value(&self) -> bool {
+        self.0 & 1 != 0
+    }
+}
+
+/// This encapsulates the bit-fields for PMU stop_flags parameter as described in the SBI documentation
+/// for sbi_pmu_counter_stop
+#[derive(Copy, Clone, Debug, Default)]
+pub struct PmuCounterStopFlags(u64);
+
+impl PmuCounterStopFlags {
+    /// Constructs a new PmuCounterStopFlags from a valid passed-in value.
+    pub fn from_raw_value(value: u64) -> Result<Self> {
+        match value {
+            0 | 1 => Ok(PmuCounterStopFlags(value)),
+            _ => Err(Error::InvalidParam),
+        }
+    }
+
+    /// Returns the raw value of the inner bit-flag field.
+    pub fn raw(&self) -> u64 {
+        self.0
+    }
+
+    /// Sets the stop_reset bit-flag (resets the counter after stopping).
+    pub fn set_reset_flag(self) -> Self {
+        PmuCounterStopFlags(self.0 | 1)
+    }
+
+    /// Returns if the stop_reset bit flag is set.
+    pub fn get_reset_flag(&self) -> bool {
+        self.0 & 1 != 0
+    }
+}
+
+/// This encapsulates the counter information returned by the call to sbi_pmu_counter_get_info.
+#[derive(Copy, Clone, Debug)]
+pub struct PmuCounterInfo(u64);
+
+impl PmuCounterInfo {
+    /// Constructs a PmuCounterInfo from the passed in value.
+    pub fn new(value: u64) -> Self {
+        PmuCounterInfo(value)
+    }
+
+    /// Returns the inner value.
+    pub fn raw(&self) -> u64 {
+        self.0
+    }
+
+    /// Returns if the counter is a hardware counter that uses the supported CSR range.
+    pub fn is_hardware_counter(&self) -> bool {
+        const CSR_CYCLE: u64 = 0xC00;
+        const CSR_HPMCOUNTER31: u64 = 0xC1F;
```

Comment:
Why CSRs are defined in the source file? Shouldn't be in some common header file ?